src/suites/cts/validation/createBindGroupLayout.spec.ts

```diff
@@ -111,17 +107,121 @@ g.test('dynamic set to true is allowed only for buffers', async t => {
+          for (const extraVisibility of shaderStages) {
+            const extraVisibilityVertex = (extraVisibility & C.ShaderStage.Vertex) !== 0;
+            if (bindingTypeInfo[extraType].isStorageBuffer && extraVisibilityVertex) continue;
+
+            kCasesForMaxResourcesPerStageTests.push({
+              maxedType,
+              maxedVisibility,
+              extraType,
+              extraVisibility,
+            });
+          }
+        }
+      }
+    }
+  }
+
+  // Should never fail unless kMaxBindingsPerBindGroup is exceeded, because the validation for
+  // resources-of-type-per-stage is in pipeline layout creation.
+  g.test('max resources per stage/in bind group layout', async t => {
+    const maxedType: GPUBindingType = t.params.maxedType;
+    const extraType: GPUBindingType = t.params.extraType;
+    const { maxedVisibility, extraVisibility } = t.params;
+    const maxedCount = bindingTypeInfo[maxedType].maxPerShaderStage;
+
+    const maxResourceBindings: GPUBindGroupLayoutBinding[] = [];
+    for (let i = 0; i < maxedCount; i++) {
+      maxResourceBindings.push({
+        binding: i,
+        visibility: maxedVisibility,
+        type: maxedType,
+      });
+    }
+
+    const goodDescriptor = { bindings: maxResourceBindings };
+
+    // Control
+    t.device.createBindGroupLayout(goodDescriptor);
+
+    const newDescriptor = clone(goodDescriptor);
+    newDescriptor.bindings.push({
+      binding: maxedCount,
+      visibility: extraVisibility,
+      type: extraType,
+    });
+
+    const shouldError = maxedCount >= kMaxBindingsPerBindGroup;
+
+    t.expectValidationError(() => {
+      t.device.createBindGroupLayout(newDescriptor);
+    }, shouldError);
+  }).params(kCasesForMaxResourcesPerStageTests);
+
+  // One pipeline layout can have a maximum number of each type of binding *per stage* (which is
+  // different for each type). Test that the max works, then add one more binding of same-or-different
+  // type and same-or-different visibility.
+  g.test('max resources per stage/in pipeline layout', async t => {
+    const maxedType: GPUBindingType = t.params.maxedType;
+    const extraType: GPUBindingType = t.params.extraType;
+    const { maxedVisibility, extraVisibility } = t.params;
+    const maxedCount = bindingTypeInfo[maxedType].maxPerShaderStage;
+
+    const maxResourceBindings: GPUBindGroupLayoutBinding[] = [];
+    for (let i = 0; i < maxedCount; i++) {
+      maxResourceBindings.push({
+        binding: i,
+        visibility: maxedVisibility,
+        type: maxedType,
+      });
+    }
+
+    const goodLayout = t.device.createBindGroupLayout({ bindings: maxResourceBindings });
+
+    // Control
+    t.device.createPipelineLayout({ bindGroupLayouts: [goodLayout] });
+
+    const extraLayout = t.device.createBindGroupLayout({
+      bindings: [{ binding: 0, visibility: extraVisibility, type: extraType }],
+    });
+
+    const hasCollision =
```

Comment:
nit: Comment to explain what a "collision" is and why it's an error, and why they must be storage buffers to be regarded as colliding.